apps/api/src/notifications/providers/notifications.sendgrid.provider.ts

```diff
@@ -39,6 +56,78 @@ export class SendGridNotificationsProvider
     return listId
   }
 
+  private async createContactExport(listId: string) {
+    const request = {
+      url: `/v3/marketing/contacts/exports`,
+      method: 'POST',
+      body: {
+        list_ids: [listId],
+        file_type: 'json',
+      },
+    } as ClientRequest
+    const [response] = await sgClient.request(request).catch((err) => {
+      throw new BadRequestException(`Couldn't create export. Error is ${err}`)
+    })
+    return response.body as { id: string }
+  }
+
+  private async getContactExportStatus(jobId: string) {
+    const request = {
+      url: `/v3/marketing/contacts/exports/${jobId}`,
+      method: 'GET',
+    } as ClientRequest
+    const [response] = await sgClient.request(request).catch((err) => {
+      throw new BadRequestException(`Couldn't create export. Error is ${err}`)
+    })
+    return response.body as SendGridExportStatusResponse
+  }
+
+  async getContactsFromList({ listId }: ContactsFromListParams) {
+    const SENDGRID_EXPORT_TIMEOUT = 10000
+    const RETRY_LIMIT = 5
+    let numOfRetries = 0
+    Logger.debug('Creating contacts exports')
+    const createContactExport = await this.createContactExport(listId)
+    const jobId = createContactExport.id
+    Logger.debug(`Created export with id ${jobId}`)
+    let exportStatusResponse = await this.getContactExportStatus(jobId)
+
+    do {
+      Logger.debug('Waiting export to be finished')
+      await new Promise((r) => setTimeout(r, SENDGRID_EXPORT_TIMEOUT))
+      exportStatusResponse = await this.getContactExportStatus(jobId)
+      Logger.debug(`Export finished with status ${exportStatusResponse.status}`)
+      switch (exportStatusResponse.status) {
+        case 'failure':
+          return Promise.reject(exportStatusResponse.message)
+        case 'ready':
+          break
+        default:
+      }
+      numOfRetries++
+    } while (exportStatusResponse.status === 'pending' && numOfRetries < RETRY_LIMIT)
+    if (numOfRetries >= RETRY_LIMIT) {
+      throw new InternalServerErrorException(
+        `Couldn't export contacts within the limit. Try again later.`,
+      )
+    }
+    const exportUrl = exportStatusResponse.urls[0]
```

Comment:
can we check if the urls list is non empty?

Reply:
Will do, though if export status is succeeded it always returns an url.